EventFilter/HcalRawToDigi/plugins/alpaka/HcalDigisProducerPortable.cc

```diff
@@ -0,0 +1,200 @@
+    device::EDPutToken<DeviceCollectionPhase1> digisF3HBToken_;
+
+    struct ConfigParameters {
+      uint32_t maxChannelsF01HE, maxChannelsF5HB, maxChannelsF3HB;
+    };
+    ConfigParameters config_;
+  };
+
+  void HcalDigisProducerPortable::fillDescriptions(edm::ConfigurationDescriptions& confDesc) {
+    edm::ParameterSetDescription desc;
+
+    // FIXME
+    desc.add<edm::InputTag>("hbheDigisLabel", edm::InputTag("hcalDigis"));
+    desc.add<edm::InputTag>("qie11DigiLabel", edm::InputTag("hcalDigis"));
+    desc.add<std::string>("digisLabelF01HE", std::string{"f01HEDigisGPU"});
+    desc.add<std::string>("digisLabelF5HB", std::string{"f5HBDigisGPU"});
+    desc.add<std::string>("digisLabelF3HB", std::string{"f3HBDigisGPU"});
+    desc.add<uint32_t>("maxChannelsF01HE", 10000u);
+    desc.add<uint32_t>("maxChannelsF5HB", 10000u);
+    desc.add<uint32_t>("maxChannelsF3HB", 10000u);
+
+    confDesc.addWithDefaultLabel(desc);
+  }
+
+  HcalDigisProducerPortable::HcalDigisProducerPortable(const edm::ParameterSet& ps)
+      : hbheDigiToken_{consumes<HBHEDigiCollection>(ps.getParameter<edm::InputTag>("hbheDigisLabel"))},
+        qie11DigiToken_{consumes<QIE11DigiCollection>(ps.getParameter<edm::InputTag>("qie11DigiLabel"))},
+        digisF01HEToken_{produces(ps.getParameter<std::string>("digisLabelF01HE"))},
+        digisF5HBToken_{produces(ps.getParameter<std::string>("digisLabelF5HB"))},
+        digisF3HBToken_{produces(ps.getParameter<std::string>("digisLabelF3HB"))} {
+    config_.maxChannelsF01HE = ps.getParameter<uint32_t>("maxChannelsF01HE");
+    config_.maxChannelsF5HB = ps.getParameter<uint32_t>("maxChannelsF5HB");
+    config_.maxChannelsF3HB = ps.getParameter<uint32_t>("maxChannelsF3HB");
+  }
+
+  void HcalDigisProducerPortable::produce(device::Event& event, device::EventSetup const& setup) {
+    const auto hbheDigis = event.getHandle(hbheDigiToken_);
+    const auto qie11Digis = event.getHandle(qie11DigiToken_);
+
+    //Get the number of samples in data from the first digi
+    auto const stride = HBHEDataFrame::MAXSAMPLES * 0.5 + 1;
+    auto const size = hbheDigis->size() * stride;  // number of channels * stride
+
+    // stack host memory in the queue
+    HostCollectionPhase0 hf5_(size, event.queue());
+
+    // device product
+    DeviceCollectionPhase0 df5_(size, alpaka::getDev(event.queue()));
+
+    // set SoA_Scalar;
+    hf5_.view().stride() = stride;
+    hf5_.view().size() = hbheDigis->size();
+
+    for (unsigned int i = 0; i < hbheDigis->size(); ++i) {
+      auto const& hbhe = (*hbheDigis)[i];
+      auto const id = hbhe.id().rawId();
+      auto const presamples = hbhe.presamples();
+      uint16_t header_word = (1 << 15) | (0x5 << 12) | (0 << 10) | ((hbhe.sample(0).capid() & 0x3) << 8);
+
+      auto hf5_vi = hf5_.view()[i];
+      hf5_vi.ids() = id;
+      hf5_vi.npresamples() = presamples;
+      hf5_vi.data()[0] = header_word;
+      //TODO:: get HEADER_WORDS/WORDS_PER_SAMPLE from DataFormat
```

Comment:
Is this still to be done?

Reply:
Unfortunately yes for flavor 5 `HEADER_WORDS`, I put them in `HcalDigiDeviceCollection.h` because these are mostly used in `Mahi.dev.cc`
https://github.com/cms-sw/cmssw/pull/44910/files#diff-28f8b5fd22ee29cd1c90a5745685aa58e5000ea01716ba00685cf91caf78b48fR47